clustermesh: minor cleanups for clustermesh metrics

MrFreezeex opened the pull request:
See commit for more details

giorio94 commented on pkg/metrics/metrics.go:
```diff
@@ -71,6 +71,9 @@ const (
 	SubsystemAPILimiter = "api_limiter"
 
+	// SubsystemClustermesh is the subsystem to scope metrics related to the clustermesh area.
+	SubsystemClustermesh = "clustermesh"
```
Nit.

```suggestion
	SubsystemClusterMesh = "clustermesh"
```

giorio94 commented on Documentation/observability/metrics.rst:
```diff
@@ -1006,25 +1006,29 @@ Name                                                 Labels
+============================================== ============================================================ ========== ===========================================================
+Name                                           Labels                                                       Default    Description
+============================================== ============================================================ ========== ===========================================================
+``clustermesh_serviceexport_info``             ``serviceexport``, ``namespace``                             Enabled    Information about ServiceExport in the local cluster
```
Should these be under the `clustermesh` or `mcsapi` subsystem? The current organization in the docs seems to imply the latter.

MrFreezeex replied:
Good point I changed those to mcsapi now :+1:

giorio94 commented on Documentation/observability/metrics.rst:
```diff
@@ -1006,25 +1006,29 @@ Name                                                 Labels
+=============================================== ============================================================ ========== ==================================================================
+Name                                            Labels                                                       Default    Description
+=============================================== ============================================================ ========== ==================================================================
+``clustermesh_remote_clusters``                 ``source_cluster``, ``source_node_name``                     Enabled    The total number of remote clusters meshed with the local cluster
```
Nit. The update of the labels should be in a separate commit.

Tangentially related, I'm personally not convinced that having a `source_node_name` label in the clustermesh metrics is actually a good idea (and the `source_cluster` one FWIW), given that they are fixed for each node, and they could be inferred differently. Additionally, no other metric actually use them (besides the health ones apparently). But that's the usual tradeoff of whether the improvement warrants a breaking change, and the possible associated pain for users.

MrFreezeex replied:
Hmm yep true, given that there's already some breaking change to clustermesh metrics this release it might (arguably) be the best time to do that.

I am preparing this change but I think I would prefer doing that in a separate PR though as it's slightly more involved to clean up everything correctly :sweat_smile:.

giorio94 commented on Documentation/operations/upgrade.rst:
```diff
@@ -395,6 +395,9 @@ now report per cluster metric instead of a "global" count and were renamed to re
 * ``cilium_operator_clustermesh_remote_cluster_services``
 * ``cilium_operator_clustermesh_remote_cluster_service_exports``
 
+* kvstoremesh clustermesh metrics are now correctly prefixed by ``cilium_kvstoremesh_clustermesh`` instead of ``cilium_``.
```
```suggestion
* kvstoremesh clustermesh metrics are now correctly prefixed by ``cilium_kvstoremesh_clustermesh`` instead of ``cilium_kvstoremesh``.
```

MrFreezeex replied:
Hmm from what I can see from the code it is indeed `cilium_` at least for the common metrics in current main

MrFreezeex replied:
(nvm I realized what was happening on your other comment here: https://github.com/cilium/cilium/pull/41924#discussion_r2390193646)

giorio94 commented on Documentation/observability/metrics.rst:
```diff
@@ -1552,17 +1556,18 @@ Name                                     Labels
 
-Remote clusters
-~~~~~~~~~~~~~~~
+Clustermesh
```
> I am guessing that the kvstoremesh metrics not including the word "clustermesh" might have been on purpose as kvstoremesh is necessarily clustermesh related [...]

Yep, that was intentional to avoid `cilium_kvstoremesh_clustermesh_remote_clusters(_...)`, which feels a bit redundant (and fairly long) to me. I don't have a too strong opinion (even though I probably slightly prefer the current form), but I'd suggest weighing in whether the breaking change is actually worthy or not.

MrFreezeex replied:
Rolled those changes back, I am still proposing to keep the Clustermesh section name in this documentation instead of remote clusters though, wdyt?

giorio94 replied:
Yeah, sounds reasonable, no hard feeling on my side.

giorio94 commented on pkg/clustermesh/kvstoremesh/metrics.go:
```diff
@@ -10,16 +10,16 @@ import (
-			Subsystem: "",
+		LeaderElectionStatus: metric.NewGauge(metric.GaugeOpts{
+			Namespace: metrics.CiliumKVStoreMeshNamespace,
+			Subsystem: metrics.SubsystemClustermesh,
```
Hmm, I wouldn't include this one under that the `clustermesh` namespace, as kvstoremesh specific. `metrics.Namespace` is also already `metrics.CiliumKVStoreMeshNamespace` in the context of KVStoreMesh, so that is a no-op change.

MrFreezeex replied:
Ahh I didn't know `metrics.Namespace` was overridden :man_facepalming:, this makes more sense to me now!

giorio94 replied:
Yep, exactly, that is a variable on purpose to allow overwriting it from other components.

giorio94 commented on pkg/clustermesh/kvstoremesh/metrics.go:
```diff
@@ -10,16 +10,16 @@ import (
+			Subsystem: metrics.SubsystemClustermesh,
 			Name:      "leader_election_master_status",
-			Help:      "leader election status",
-		}, []string{metrics.LabelLeaderElectionName}),
```
The goal here was to mimic the corresponding controller-manager metric: https://github.com/cilium/cilium/pull/39848#discussion_r2231103649. I don't have a strong opinion though. (/cc @balous FYI)

MrFreezeex replied:
Ah fair enough, I didn't noticed this in the original PR, will change that back and add some comment!

giorio94 reviewed (commented):
Thanks! A couple of comments inline.

MrFreezeex committed:
clustermesh: add common clustermesh subsystem variable

Also remove `ConfigName` that are now automatically computed

Signed-off-by: Arthur Outhenin-Chalandre <git@mrfreezeex.fr>

MrFreezeex committed:
clustermesh: add mcsapi subsystem to metrics

Signed-off-by: Arthur Outhenin-Chalandre <git@mrfreezeex.fr>

MrFreezeex committed:
metrics: document that the Namespace metrics can change at runtime

Signed-off-by: Arthur Outhenin-Chalandre <git@mrfreezeex.fr>

MrFreezeex committed:
clustermesh: add missing metrics in documentation

Signed-off-by: Arthur Outhenin-Chalandre <git@mrfreezeex.fr>

MrFreezeex committed:
clustermesh: baked-in the name label in kvstoremesh election metric

This should improve readability as we can define the label value on its
definition and explain what is happening too in a comment.

Signed-off-by: Arthur Outhenin-Chalandre <git@mrfreezeex.fr>

giorio94 commented on pkg/clustermesh/metrics.go:
```diff
@@ -22,25 +22,25 @@ type Metrics struct {
 	return Metrics{
 		TotalNodes: metric.NewGaugeVec(metric.GaugeOpts{
-			ConfigName: metrics.Namespace + "_" + subsystem + "_remote_cluster_nodes",
+			ConfigName: metrics.Namespace + "_" + metrics.SubsystemClusterMesh + "_remote_cluster_nodes",
```
While you are here, could you drop the `ConfigName` field? It used to be necessary back then, but it is now [automatically computed if unset](https://github.com/cilium/cilium/blob/5ce5669732e96dc41dc02ec64ef66c1cd3f84fef/pkg/metrics/metric/metric.go#L275-L280).

giorio94 commented on pkg/metrics/metrics.go:
```diff
@@ -258,8 +258,9 @@ var (
-	// Namespace is used to scope metrics from cilium. It is prepended to metric
-	// names and separated with a '_'
+	// Namespace is used to scope metrics from the current cilium component by
+	// overwriting its value at runtime. It is prepended to metric names and
```
Nit. The caveat is that it can only be changed very early during initialization, before actually starting the hive, otherwise it would not have any effect.

MrFreezeex replied:
Thanks! I tried adding a little clarification that this is done before hive start

giorio94 commented on Documentation/observability/metrics.rst:
```diff
@@ -1552,17 +1556,20 @@ Name                                     Labels
+``remote_cluster_failures``                     ``source_cluster``, ``source_node_name``, ``target_cluster`` The total number of failures related to the remote cluster
+``remote_cluster_last_failure_ts``              ``source_cluster``, ``source_node_name``, ``target_cluster`` The timestamp of the last failure of the remote cluster
+``remote_cluster_readiness_status``             ``source_cluster``, ``source_node_name``, ``target_cluster`` The readiness status of the remote cluster
+``leader_election_master_status``               ``name``                                                     The leader election status
```
Nit. The last one should be in a separate category, as not related to the clustermesh subsystem.

MrFreezeex replied:
I added a dedicated KVStoremesh category

giorio94 commented on pkg/clustermesh/kvstoremesh/metrics.go:
```diff
@@ -10,16 +10,18 @@ import (
-			Help:      "leader election status",
-		}, []string{metrics.LabelLeaderElectionName}),
+			Help:      "The leader election status",
+		}, []string{"name"}).WithLabelValues("kvstoremesh"),
```
Nit. You could probably use `ConstLabels` to achieve the same result. The downside of both these approaches is that it would not be possible to reuse the metric for a different leader election in case we needed to introduce one (although I recon that it feels fairly unlikely).

MrFreezeex replied:
Ah TIL, adding const labels then!

And yes as soon as we have a second one in kvstoremesh (if we really end up doing that :sweat_smile:) the explanation of why we are forcing a label name no longer stands so should be fine to just revert this back to the way it was before ~ :+1:

giorio94 reviewed (approved):
Thanks!
